### ISSUE TITLE
Confusing behavior with inheritance and *args

### ISSUE THREAD

**Issue by danielgordon10** [microsoft/pyright#1556]
I don't know what the "correct" behavior is in this case, but the current behavior is a bit confusing and may result in a lot of false positives. I think it's pretty common pattern to have something like the following.
```python
import abc
class SuperClass(abc.ABC):
    @abc.abstractmethod
    def func(self, *args, **kwargs):
        raise NotImplementedError

class SubClass(SuperClass):
    def func(self, actual_argument: int):
        return 2 * actual_argument
```

This results in numerous errors:
```
  8:9 - error: Method "func" overrides class "SuperClass" in an incompatible manner
    Parameter count mismatch; base method has 3, but override has 2
    Parameter 2 type mismatch: base parameter is type "Unknown", override parameter is type "int" (reportIncompatibleMethodOverride)
```

Even trying to make the code a little more "similar" between the super and sub functions doesn't help (and gives an even more confusing error message).
```python
import abc
from typing import Any
class SuperClass(abc.ABC):
    @abc.abstractmethod
    def func(self, *args: Any):
        raise NotImplementedError

class SubClass(SuperClass):
    def func(self, actual_argument: Any):
        return 2 * actual_argument
```

Error:
```
  9:9 - error: Method "func" overrides class "SuperClass" in an incompatible manner
    Parameter 2 type mismatch: base parameter is type "Any", override parameter is type "Any"
```

The only way to not get errors would be to have the sub function have exactly the same signature.

Based on the principles of inheritance, one could argue that the subclass func ideally should always match the signature of the super func to retain invariances. However, in practice since the subclass func doesn't ever call super, it may actually have a more restrictive signature which is often reflected in practice.

**Comment by erictraut** [comment]
This error is produced only if you enable the `reportIncompatibleMethodOverride` diagnostic check. This check is specifically meant to detect unsafe situations with respect to overrides. Your example above is unsafe because a function that receives an object of type `SuperClass` could attempt to call `func` with any argument list, but this would cause a runtime error if the object were a `SubClass` instance. This is the very problem that this diagnostic check is designed to catch and prevent.

While this may be a common pattern, it's a very unsafe practice, and I'd strongly recommend against it if you want your code to be as safe and robust as possible.

You have two options:
1. Disable the `reportIncompatibleMethodOverride` diagnostic check if you don't care about detecting this class of errors.
2. Fix your code so the the signatures match between the parent class and subclass.

**Comment by danielgordon10** [comment]
Right. I know I can disable it but then I'm afraid of it missing some actual cases of mistaken function signatures. But I guess what you're saying is that would be caught by other checks? That's certainly possible if the objects are resolved to the most specific class they can be.
